### PR TITLE
uc-crux-llvm: Improve representation of LLVM types

### PR DESCRIPTION
**Pull request by langston-barrett**
For opaque pointers, replaces value-level representation of type names with value-and-type level representation with a `SymbolRepr`.

For structs, represents packed-ness at the type-level and remove `StructInfo` field. This hs three nice effects:

* Makes `FullTypeRepr` into a true singleton, making its `TestEquality` instances lawful.
* Reflects the fact that packed and non-packed structs are not the same at the type level.
* Reduces redundant storage, instead recomputing `StructInfo`s on demand.

Also introduces a wrapper around Crucible's `DataLayout` with an additional phantom type parameter that marks it as corresponding to a particular LLVM module.

### CODE REVIEW

File: uc-crux-llvm/src/UCCrux/LLVM/Classify.hs [hunk 1]
```diff
@@ -123,14 +120,13 @@ classifyAssertion _sym predicate loc =
...
-  DataLayout ->
   ModuleTypes m ->
   What4.ConcreteVal (What4.BaseBVType w) ->
   PartTypeRepr m ft ->
   Int
-elemsFromOffset dl mts offset partType =
+elemsFromOffset mts offset partType =
   let pointedTo = asFullType mts partType
-      typeSize = bytesToInteger (memTypeSize dl (toMemType pointedTo))
+      typeSize = sizeInBytes mts pointedTo 1 -- 1 = array length
```
Comment (travitch) [inline uc-crux-llvm/src/UCCrux/LLVM/Classify.hs:129]:
```suggestion
      singletonArrayLength = 1
      typeSize = sizeInBytes mts pointedTo singletonArrayLength
```

File: uc-crux-llvm/src/UCCrux/LLVM/FullType/Type.hs [hunk 2]
```diff
@@ -418,7 +472,9 @@ toFullTypeM memType =
               let ?lc = typeContext mts
               Some ftRepr <-
                 case asMemType' strIdent of
-                  Left opaqueIdent -> pure $ Some (FTOpaquePtrRepr opaqueIdent)
+                  Left (L.Ident opaqueIdent) ->
+                    case someSymbol (Text.pack opaqueIdent) of
+                      Some symRep -> pure $ Some (FTOpaquePtrRepr symRep)
```
Comment (travitch) [inline uc-crux-llvm/src/UCCrux/LLVM/FullType/Type.hs:477]: Is the opaque pointer identifier meaningful here? Are they really distinct nominal types, or are there cases where opaque pointers get fudged?

I guess I'm trying to understand what the symbol is representing, what tradeoffs that implies, and if that is actually a safe/desirable thing to represent here.

File: uc-crux-llvm/src/UCCrux/LLVM/FullType/MemType.hs [hunk 3]
```diff
@@ -24,44 +24,52 @@ import           Data.Parameterized.TraversableFC (toListFC)
...
+    FTUnboundedArrayRepr fullTypeRepr -> ArrayType 0 (toMemType dl fullTypeRepr)
+    FTStructRepr sp fields ->
+      let cdl = crucibleDataLayout dl
+          memFields = toListFC (toMemType dl) fields
+      in StructType (mkStructInfo cdl (structPackedReprToBool sp) memFields)
   where
-    funType :: Maybe (FullTypeRepr m ft) -> Ctx.Assignment (FullTypeRepr m) argTypes -> Bool -> MemType
-    funType maybeRetRepr argsRepr isVarArgs =
+    funType ::
+      DataLayout m ->
```
Comment (travitch) [inline uc-crux-llvm/src/UCCrux/LLVM/FullType/MemType.hs:62]: Is this argument necessary? It looks like this function is only called in two places that just pass the (already visible) `dl` to this function

File: uc-crux-llvm/src/UCCrux/LLVM/Classify.hs (hunk 1 above)
Comment (langston-barrett) [reply uc-crux-llvm/src/UCCrux/LLVM/Classify.hs]: Thanks for the comment, I've refactored the module containing this helper to have array and non-array helpers, which also clearified the code in a few other places.

File: uc-crux-llvm/src/UCCrux/LLVM/FullType/MemType.hs (hunk 3 above)
Comment (langston-barrett) [reply uc-crux-llvm/src/UCCrux/LLVM/FullType/MemType.hs:62]: It's a trade-off between using `ScopedTypeVariables` to match the `m` in this signature with the `m` in the outer signature vs just passing it in here. Happy to change it to used `ScopedTypedVariables` if you prefer it!

Comment (travitch) [reply uc-crux-llvm/src/UCCrux/LLVM/FullType/MemType.hs:62]: ah, I had assumed scoped type variables was already enabled. I think that using scoped type variables might be less confusing in the long run, as there is no possibility of confusing values.

**Review by travitch — APPROVED** [review]

**Pull request merged**